flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationTestBase.java

```diff
@@ -0,0 +1,569 @@
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.enableCheckpointing(checkpointInterval)
+                .enableChangelogStateBackend(true)
+                .getCheckpointConfig()
+                .setCheckpointStorage(checkpointFile.toURI());
+        env.setStateBackend(stateBackend)
+                .setRestartStrategy(RestartStrategies.fixedDelayRestart(restartAttempts, 0));
+        return env;
+    }
+
+    protected StreamExecutionEnvironment getEnv(StateBackend stateBackend) throws IOException {
+        return getEnv(stateBackend, TEMPORARY_FOLDER.newFolder(), 10, 1);
+    }
+
+    protected JobGraph buildJobGraph(
+            StreamExecutionEnvironment env, ControlledSource controlledSource, JobID jobId) {
+        env.addSource(controlledSource)
+                .keyBy(element -> element)
+                .map(new CountMapper())
+                .addSink(new CollectionSink())
+                .setParallelism(1);
+        return env.getStreamGraph().getJobGraph(jobId);
+    }
+
+    protected void waitAndAssert(JobGraph jobGraph) throws Exception {
+        waitUntilJobFinished(jobGraph);
+        assertEquals(CollectionSink.getActualResult(), ControlledSource.getExpectedResult());
+    }
+
+    protected JobID generateJobID() {
+        byte[] randomBytes = new byte[AbstractID.SIZE];
+        ThreadLocalRandom.current().nextBytes(randomBytes);
+        return JobID.fromByteArray(randomBytes);
+    }
+
+    protected static List<Long> getAllMaterializationId(File checkpointFile, JobID jobID)
+            throws IOException {
+        List<Path> checkpointPaths = findAllSortedExternalizedCheckpoint(checkpointFile, jobID);
+        for (Path checkpointPath : checkpointPaths) {
+            CheckpointMetadata checkpointMetadata =
+                    loadCheckpointMetadata(checkpointPath.toString());
+            List<Long> materializationIds =
+                    checkpointMetadata.getOperatorStates().stream()
+                            .flatMap(operatorState -> operatorState.getStates().stream())
+                            .flatMap(
+                                    operatorSubtaskState ->
+                                            operatorSubtaskState.getManagedKeyedState().stream())
+                            .map(keyedStateHandle -> (ChangelogStateBackendHandle) keyedStateHandle)
+                            .filter(
+                                    changelogStateBackendHandle ->
+                                            !changelogStateBackendHandle
+                                                    .getMaterializedStateHandles()
+                                                    .isEmpty())
+                            .map(ChangelogStateBackendHandle::getMaterializationID)
+                            .collect(Collectors.toList());
+            if (!materializationIds.isEmpty()) {
+                return materializationIds;
+            }
```

Comment:
This means the 1st checkpoint with some materialized state will be returned, right?
Is there any particular reason why not the last one?
I think at least in case of failover, it could break the check that there were materialization after the recovery

Reply:
Replace `findExternalizedCheckpoint` with `getMostRecentCompletedCheckpointMaybe`